pkg/messageprocessors/normalizedpayload/uplink.go

```diff
@@ -0,0 +1,235 @@
+		}
+		n := val.NumberValue
+		if err := validate(n, vals, path); err != nil {
+			return err
+		}
+		*selector(dst) = &n
+		return nil
+	}
+}
+
+// minimum returns a field validator that checks the inclusive minimum.
+func minimum[T constraints.Ordered](min T) fieldValidator[T] {
+	return func(v T, path string) error {
+		if v < min {
+			return errFieldMinimum.WithAttributes(
+				"path", path,
+				"minimum", min,
+			)
+		}
+		return nil
+	}
+}
+
+// exclusiveMinimum returns a field validator that checks the exclusive minimum.
+//nolint:unused,deadcode
+func exclusiveMinimum[T constraints.Ordered](min T) fieldValidator[T] {
+	return func(v T, path string) error {
+		if v <= min {
+			return errFieldExclusiveMinimum.WithAttributes(
+				"path", path,
+				"minimum", min,
+			)
+		}
+		return nil
+	}
+}
+
+// maximum returns a field validator that checks the inclusive maximum.
+func maximum[T constraints.Ordered](max T) fieldValidator[T] {
+	return func(v T, path string) error {
+		if v > max {
+			return errFieldMaximum.WithAttributes(
+				"path", path,
+				"maximum", max,
+			)
+		}
+		return nil
+	}
+}
+
+// exclusiveMaximum returns a field validator that checks the exclusive maximum.
+func exclusiveMaximum[T constraints.Ordered](max T) fieldValidator[T] {
+	return func(v T, path string) error {
+		if v >= max {
+			return errFieldExclusiveMaximum.WithAttributes(
+				"path", path,
+				"maximum", max,
+			)
+		}
+		return nil
+	}
+}
+
+var fieldParsers = map[string]fieldParser{
+	"time": parseTime(func(dst *Measurement) **time.Time { return &dst.Time }),
+	"air":  object(func(dst *Measurement) **Air { return &dst.Air }),
+	"air.temperature": parseNumber(func(dst *Measurement) **float64 { return &dst.Air.Temperature },
+		minimum(-273.15),
+	),
+	"air.relativeHumidity": parseNumber(func(dst *Measurement) **float64 { return &dst.Air.RelativeHumidity },
+		minimum(0.0),
+		maximum(1.0),
+	),
+	"air.pressure": parseNumber(func(dst *Measurement) **float64 { return &dst.Air.Pressure },
+		minimum(900.0),
+		maximum(1100.0),
+	),
+	"wind": object(func(dst *Measurement) **Wind { return &dst.Wind }),
+	"wind.speed": parseNumber(func(dst *Measurement) **float64 { return &dst.Wind.Speed },
+		minimum(0.0),
+	),
+	"wind.direction": parseNumber(func(dst *Measurement) **float64 { return &dst.Wind.Direction },
+		minimum(0.0),
+		exclusiveMaximum(360.0),
+	),
+}
+
+// Parse parses and validates the measurements.
+func Parse(measurements []*pbtypes.Struct) ([]Measurement, error) {
+	res := make([]Measurement, len(measurements))
+	for i, src := range measurements {
+		err := parse(&res[i], src, "")
+		if err != nil {
+			return nil, err
+		}
+	}
+	return res, nil
+}
+
+func parse(dst *Measurement, src *pbtypes.Struct, prefix string) error {
+	for k, v := range src.GetFields() {
+		path := fmt.Sprintf("%s%s", prefix, k)
+		parser, ok := fieldParsers[path]
+		if !ok {
+			return errUnknownField.WithAttributes("path", path)
+		}
```

Comment:
What basically happens is that because [`additionalProperties` aren't allowed](https://github.com/TheThingsNetwork/lorawan-devices/pull/508/files#diff-9fcde326d127f74194f70e563bdf2c118c51b719c308f015b8eb0204a9a552fbR296), if any field is present that is not a known normalized field, there's an immediate error (2/3)